xarray/core/variable.py

```diff
@@ -384,34 +396,174 @@ def _item_key_to_tuple(self, key):
+        dims = []
+        for k, d in zip(key, self.dims):
+            if isinstance(k, Variable):
+                if len(k.dims) > 1:
+                    return self._broadcast_indexes_vectorized(key)
+                dims.append(k.dims[0])
+            if not isinstance(k, integer_types):
+                dims.append(d)
+
+        if len(set(dims)) == len(dims):
+            return self._broadcast_indexes_outer(key)
+
+        return self._broadcast_indexes_vectorized(key)
+
+    def _broadcast_indexes_basic(self, key):
+        dims = tuple(dim for k, dim in zip(key, self.dims)
+                     if not isinstance(k, integer_types))
+        return dims, BasicIndexer(key), None
+
+    def _broadcast_indexes_outer(self, key):
+        dims = tuple(k.dims[0] if isinstance(k, Variable) else dim
+                     for k, dim in zip(key, self.dims)
+                     if not isinstance(k, integer_types))
+        indexer = []
+        for k in key:
+            if isinstance(k, Variable):
+                k = k.data
+
+            if isinstance(k, BASIC_INDEXING_TYPES):
+                indexer.append(k)
+            else:
+                k = np.asarray(k)
+                if k.ndim > 1:
+                    raise IndexError("Unlabeled multi-dimensional array "
+                                     "cannot be used for indexing: {}".format(
+                                         k))
+                indexer.append(k if k.dtype.kind != 'b' else np.flatnonzero(k))
+        return dims, OuterIndexer(indexer), None
+
+    def _nonzero(self):
+        """ Equivalent numpy's nonzero but returns a tuple of Varibles. """
+        # TODO we should replace dask's native nonzero
+        # after https://github.com/dask/dask/issues/1076 is implemented.
+        nonzeros = np.nonzero(self.data)
+        return tuple(Variable((dim), nz) for nz, dim
+                     in zip(nonzeros, self.dims))
+
+    def _broadcast_indexes_vectorized(self, key):
+
+        variables = []
+        out_dims_set = OrderedSet()
+        for dim, value in zip(self.dims, key):
+            if isinstance(value, slice):
+                out_dims_set.add(dim)
+            else:
+                try:
+                    variable = (value if isinstance(value, Variable) else
+                                as_variable(value, name=dim))
+                except MissingDimensionsError:  # change to better exception
+                    raise IndexError("Unlabeled multi-dimensional array "
+                                     "cannot be used for indexing.")
+
+                if variable.dtype.kind == 'b':  # boolean indexing case
+                    if variable.ndim > 1:
+                        raise IndexError("{}-dimensional boolean indexing is "
+                                         "not supported. ".format(
+                                                                variable.ndim))
+                    (variable,) = variable._nonzero()
```

Comment:
It occurs to that we should also have another sanity check here, to verify that the variable is the same length as dimension on the object being indexed. Cases like `xr.Variable(('x',), [1, 2, 3])[np.array([True, False])]` should be an error.

Reply:
A sanity check and tests added.